File: go/hyperliquid/prediction.go
package hyperliquid

import (
	"context"
	"fmt"
	"strings"
	"unicode"
)

func parseOutcomeDescription(description string) map[string]string {
	fields := make(map[string]string)
	for _, part := range strings.Split(description, "|") {
		pieces := strings.SplitN(part, ":", 2)
		if len(pieces) != 2 {
			continue
		}
		fields[pieces[0]] = pieces[1]
	}
	return fields
}

func formatPredictionExpiry(expiry string) string {
	if len(expiry) != 13 || expiry[8] != '-' {
		return expiry
	}
	return fmt.Sprintf("%s-%s-%sT%s:%s:00Z", expiry[:4], expiry[4:6], expiry[6:8], expiry[9:11], expiry[11:13])
}

func predictionTitle(fields map[string]string) string {
	underlying := fields["underlying"]
	if underlying == "" {
		underlying = "Outcome"
	}
	targetPrice := fields["targetPrice"]
	expiry := formatPredictionExpiry(fields["expiry"])
	if targetPrice != "" && expiry != "" {
		return fmt.Sprintf("%s above %s on %s", underlying, targetPrice, expiry)
	}
	if targetPrice != "" {
		return fmt.Sprintf("%s above %s", underlying, targetPrice)
	}
	return underlying
}

func predictionSlug(value string) string {
	var b strings.Builder
	lastDash := false
	for _, r := range strings.ToLower(value) {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			b.WriteRune(r)
			lastDash = false
		} else if !lastDash && b.Len() > 0 {
			b.WriteByte('-')
			lastDash = true
		}
	}
	return strings.Trim(b.String(), "-")
}

// appStylePredictionSlugMonths is only used to generate app-style slug aliases such as
// btc-above-78213-yes-may-04-0600 from outcomeMeta expiry 20260504-0600.
var appStylePredictionSlugMonths = map[string]string{
	"01": "jan",
	"02": "feb",
	"03": "mar",
	"04": "apr",
	"05": "may",
	"06": "jun",
	"07": "jul",
	"08": "aug",
	"09": "sep",
	"10": "oct",
	"11": "nov",
	"12": "dec",
}

func appStylePredictionSlug(fields map[string]string, side string) string {
	underlying := fields["underlying"]
	targetPrice := fields["targetPrice"]
	expiry := fields["expiry"]
	if underlying == "" || targetPrice == "" || len(expiry) != 13 {
		return ""
	}
	month := appStylePredictionSlugMonths[expiry[4:6]]
	if month == "" {
		return ""
	}
	parts := []string{underlying, "above", targetPrice}
	if side != "" {
		parts = append(parts, side)
	}
	parts = append(parts, month, expiry[6:8], expiry[9:13])
	return predictionSlug(strings.Join(parts, "-"))
}

func (p PredictionMarket) matches(query string) bool {
	normalized := strings.ToLower(query)
	values := []string{
		p.Slug,
		strings.ToLower(p.Title),
		strings.ToLower(p.Name),
		strings.ToLower(p.Underlying),
		strings.ToLower(p.Yes.Symbol),
		strings.ToLower(p.No.Symbol),
		strings.ToLower(p.Yes.Token),
		strings.ToLower(p.No.Token),
	}
	for _, alias := range p.Aliases {
		values = append(values, strings.ToLower(alias))
	}
	for _, value := range values {
		if value == normalized || strings.Contains(value, normalized) {
			return true
		}
	}
	return false
}

func assetName(asset any) string {
	switch v := asset.(type) {
	case string:
		return v
	case PredictionSide:
		return v.Symbol
	case *PredictionSide:
		if v == nil {
			return ""
		}
		return v.Symbol
	case fmt.Stringer:
		return v.String()
	default:
		return fmt.Sprintf("%v", v)
	}
}

func isPredictionAsset(asset string) bool {
	if len(asset) > 1 && (asset[0] == '#' || asset[0] == '+') {
		return allDecimalDigits(asset[1:])
	}
	if allDecimalDigits(asset) {
		return NewDecimal(asset).Float64() >= 100000000
	}
	return false
}

func predictionSymbol(asset string) string {
	if len(asset) > 1 && asset[0] == '+' {
		return "#" + asset[1:]
	}
	if allDecimalDigits(asset) {
		id := int(NewDecimal(asset).Float64())
		if id >= 100000000 {
			return fmt.Sprintf("#%d", id-100000000)
		}
	}
	return asset
}

// PredictionMarkets lists active HIP-4 prediction markets with tradeable yes/no sides.
func (s *SDK) PredictionMarkets() (PredictionMarkets, error) {
	ctx := context.Background()
	outcomeRaw, err := s.postInfo(ctx, map[string]any{"type": "outcomeMeta"})
	if err != nil {
		return nil, err
	}
	midsRaw, err := s.postInfo(ctx, map[string]any{"type": "allMids"})
	if err != nil {
		return nil, err
	}

	mids, _ := midsRaw.(map[string]any)
	outcomeMeta, _ := outcomeRaw.(map[string]any)
	outcomes, _ := outcomeMeta["outcomes"].([]any)

	markets := PredictionMarkets{}
	for _, item := range outcomes {
		outcome, ok := item.(map[string]any)
		if !ok {
			continue
		}
		outcomeID := int(NewDecimal(outcome["outcome"]).Float64())
		description, _ := outcome["description"].(string)
		fields := parseOutcomeDescription(description)
		title := predictionTitle(fields)
		sideSpecs, _ := outcome["sideSpecs"].([]any)

		sides := make([]PredictionSide, 0, len(sideSpecs))
		for sideIndex, sideItem := range sideSpecs {
			sideSpec, _ := sideItem.(map[string]any)
			encoding := outcomeID*10 + sideIndex
			symbol := fmt.Sprintf("#%d", encoding)
			mid, _ := mids[symbol].(string)
			name, _ := sideSpec["name"].(string)
			if name == "" {
				name = fmt.Sprintf("%d", sideIndex)
			}
			sides = append(sides, PredictionSide{
				Outcome:             outcomeID,
				Side:                sideIndex,
				Name:                name,
				Symbol:              symbol,
				Token:               fmt.Sprintf("+%d", encoding),
				AssetID:             100000000 + encoding,
				Mid:                 mid,
				SzDecimals:          0,
				SupportsPriorityFee: false,
			})
		}
		if len(sides) < 2 {
			continue
		}

		name, _ := outcome["name"].(string)
		slug := appStylePredictionSlug(fields, "")
		if slug == "" {
			slug = predictionSlug(title)
		}
		aliases := []string{predictionSlug(title)}
		for _, side := range sides[:2] {
			if alias := appStylePredictionSlug(fields, side.Name); alias != "" {
				aliases = append(aliases, alias)
			}
		}

		markets = append(markets, PredictionMarket{
			Outcome:       outcomeID,
			Name:          name,
			Description:   description,
			Title:         title,
			Slug:          slug,
			Underlying:    fields["underlying"],
			TargetPrice:   fields["targetPrice"],
			Expiry:        formatPredictionExpiry(fields["expiry"]),
			Period:        fields["period"],
			Collateral:    "USDH",
			MinOrderValue: "10",
			Aliases:       aliases,
			Yes:           sides[0],
			No:            sides[1],
			Sides:         sides,
		})
	}

	return markets, nil
}

// Predictions is an alias for PredictionMarkets.
func (s *SDK) Predictions() (PredictionMarkets, error) {
	return s.PredictionMarkets()
}

// PredictionMarket finds one active HIP-4 prediction market.
func (s *SDK) PredictionMarket(filter PredictionMarketFilter) (*PredictionMarket, error) {
	markets, err := s.PredictionMarkets()
	if err != nil {
		return nil, err
	}
	if market, ok := markets.Find(filter); ok {
		return market, nil
	}
	return nil, ValidationError("no matching prediction market found").
		WithGuidance("Call sdk.PredictionMarkets() to list active HIP-4 markets.")
}

// PredictionSides returns a flat list of tradeable HIP-4 sides.
func (s *SDK) PredictionSides() ([]PredictionSide, error) {
	markets, err := s.PredictionMarkets()
	if err != nil {
		return nil, err
	}
	sides := []PredictionSide{}
	for _, market := range markets {
		sides = append(sides, market.Sides...)
	}
	return sides, nil
}

// BuyUSDH buys USDH with USDC for HIP-4 prediction markets.
func (s *SDK) BuyUSDH(amountUSDC any, opts ...OrderOption) (*PlacedOrder, error) {
	opts = append(opts, WithNotional(NewDecimal(amountUSDC).Float64()), WithTIF(TIFMarket))
	return s.Buy("@230", opts...)
}

// SellUSDH sells USDH back to USDC.
func (s *SDK) SellUSDH(amountUSDH any, opts ...OrderOption) (*PlacedOrder, error) {
	opts = append(opts, WithSize(amountUSDH), WithTIF(TIFMarket))
	return s.Sell("@230", opts...)
}

File: python/hyperliquid_sdk/prediction.py
"""HIP-4 prediction market helpers."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Iterable, Optional


def parse_outcome_description(description: str) -> dict[str, str]:
    """Parse Hyperliquid's outcome description string into key/value fields."""
    fields: dict[str, str] = {}
    for part in description.split("|"):
        if ":" not in part:
            continue
        key, value = part.split(":", 1)
        fields[key] = value
    return fields


def format_expiry(expiry: Optional[str]) -> Optional[str]:
    """Convert YYYYMMDD-HHMM into a readable UTC timestamp."""
    if not expiry or len(expiry) != 13 or expiry[8] != "-":
        return expiry
    return f"{expiry[:4]}-{expiry[4:6]}-{expiry[6:8]}T{expiry[9:11]}:{expiry[11:13]}:00Z"


def title_from_fields(fields: dict[str, str]) -> str:
    underlying = fields.get("underlying", "Outcome")
    target_price = fields.get("targetPrice")
    expiry = format_expiry(fields.get("expiry"))

    if target_price and expiry:
        return f"{underlying} above {target_price} on {expiry}"
    if target_price:
        return f"{underlying} above {target_price}"
    return underlying


def slugify(value: str) -> str:
    chars = []
    for ch in value.lower():
        if ch.isalnum():
            chars.append(ch)
        elif chars and chars[-1] != "-":
            chars.append("-")
    return "".join(chars).strip("-")


# Only used to generate app-style slug aliases such as
# btc-above-78213-yes-may-04-0600 from outcomeMeta expiry 20260504-0600.
APP_STYLE_PREDICTION_SLUG_MONTHS = {
    "01": "jan",
    "02": "feb",
    "03": "mar",
    "04": "apr",
    "05": "may",
    "06": "jun",
    "07": "jul",
    "08": "aug",
    "09": "sep",
    "10": "oct",
    "11": "nov",
    "12": "dec",
}


def app_style_prediction_slug(fields: dict[str, str], side: Optional[str] = None) -> Optional[str]:
    underlying = fields.get("underlying")
    target_price = fields.get("targetPrice")
    expiry = fields.get("expiry")
    if not underlying or not target_price or not expiry or len(expiry) != 13:
        return None
    month = APP_STYLE_PREDICTION_SLUG_MONTHS.get(expiry[4:6])
    if month is None:
        return None
    parts = [underlying, "above", target_price]
    if side:
        parts.append(side)
    parts.extend([month, expiry[6:8], expiry[9:13]])
    return slugify("-".join(parts))


@dataclass(frozen=True)
class PredictionSide:
    """A tradeable HIP-4 outcome side."""

    outcome: int
    side: int
    name: str
    symbol: str
    token: str
    asset_id: int
    mid: Optional[str] = None

    @property
    def sz_decimals(self) -> int:
        return 0

    @property
    def supports_priority_fee(self) -> bool:
        return False

    def __str__(self) -> str:
        return self.symbol


@dataclass(frozen=True)
class PredictionMarket:
    """A HIP-4 prediction market with yes/no tradeable sides."""

    outcome: int
    name: str
    description: str
    title: str
    slug: str
    underlying: Optional[str]
    target_price: Optional[str]
    expiry: Optional[str]
    period: Optional[str]
    collateral: str
    min_order_value: str
    aliases: tuple[str, ...]
    yes: PredictionSide
    no: PredictionSide
    sides: tuple[PredictionSide, ...]

    def matches(self, query: str) -> bool:
        normalized = query.lower()
        values = [
            self.slug,
            self.title.lower(),
            self.name.lower(),
            self.underlying.lower() if self.underlying else "",
            self.yes.symbol.lower(),
            self.no.symbol.lower(),
            self.yes.token.lower(),
            self.no.token.lower(),
            *self.aliases,
        ]
        return normalized in values or any(normalized in value for value in values if value)


class PredictionMarkets(list[PredictionMarket]):
    """List wrapper with a small, readable finder."""

    def find(
        self,
        query: Optional[str] = None,
        *,
        underlying: Optional[str] = None,
        target_price: Optional[str] = None,
        expiry: Optional[str] = None,
    ) -> Optional[PredictionMarket]:
        for market in self:
            if query is not None and not market.matches(query):
                continue
            if underlying is not None and (market.underlying or "").lower() != underlying.lower():
                continue
            if target_price is not None and market.target_price != str(target_price):
                continue
            if expiry is not None and market.expiry not in {expiry, format_expiry(expiry)}:
                continue
            return market
        return None


def build_prediction_markets(outcomes: Iterable[dict[str, Any]], mids: dict[str, Any]) -> PredictionMarkets:
    markets = PredictionMarkets()
    for outcome in outcomes:
        outcome_id = int(outcome["outcome"])
        description = str(outcome.get("description", ""))
        fields = parse_outcome_description(description)
        title = title_from_fields(fields)
        side_specs = outcome.get("sideSpecs", [])

        sides = []
        for side_index, side_spec in enumerate(side_specs):
            encoding = outcome_id * 10 + side_index
            symbol = f"#{encoding}"
            token = f"+{encoding}"
            sides.append(
                PredictionSide(
                    outcome=outcome_id,
                    side=side_index,
                    name=str(side_spec.get("name", side_index)),
                    symbol=symbol,
                    token=token,
                    asset_id=100_000_000 + encoding,
                    mid=mids.get(symbol),
                )
            )

        if len(sides) < 2:
            continue

        slug = app_style_prediction_slug(fields) or slugify(title)
        aliases = tuple(
            alias
            for alias in (
                slugify(title),
                app_style_prediction_slug(fields, sides[0].name),
                app_style_prediction_slug(fields, sides[1].name),
            )
            if alias
        )
        markets.append(
            PredictionMarket(
                outcome=outcome_id,
                name=str(outcome.get("name", "")),
                description=description,
                title=title,
                slug=slug,
                underlying=fields.get("underlying"),
                target_price=fields.get("targetPrice"),
                expiry=format_expiry(fields.get("expiry")),
                period=fields.get("period"),
                collateral="USDH",
                min_order_value="10",
                aliases=aliases,
                yes=sides[0],
                no=sides[1],
                sides=tuple(sides),
            )
        )
    return markets
